### DISCUSSION TITLE
unbound variable error

### DISCUSSION THREAD

**Discussion by drzhangyin** [PGScatalog/pgsc_calc#106] [Q&A] [answered]
Hello, I am new to PGScatalog pipeline, and I am trying to use it to calculate PRS strictly following the instruction. 

When I tested nextflow run pgscatalog/pgsc_calc -profile test,singularity, I received the following error message: 
Command error:
  bash: line 15: __lmod_vx: unbound variable
  .command.sh: line 2:    37 Illegal instruction     combine_scorefiles -s PGS001229_22.txt -t GRCh37 -o scorefiles.txt.gz -l log_scorefiles.json -v

When I tried to calculate PRS using PGS000004 as an example, I received similar error message:
Command error:
  bash: line 15: __lmod_vx: unbound variable
  .command.sh: line 2:    36 Illegal instruction     download_scorefiles -i PGS000004 -b GRCh37 -o $PWD -v -c pgsc_calc/1.3.2 

It would be great to have your suggestions on how I can address this error.

Thanks!!

**Comment by nebfield** [comment]
Hello,

This looks like a problem with your local computing environment related to https://github.com/TACC/Lmod

Does `$ nextflow run hello` run OK?

You should see the following output:

```
Bonjour world!

Hello world!

Hola world!

Ciao world!
```

Also, please report the version of singularity you're running:

```
$ singularity --version
singularity version 3.8.7-1.el8
```

Thanks

**Reply by drzhangyin** [reply]
Thanks for the feedback!
The $ nextflow run hello runs well, and I see one of those four outputs each time.
The version of singularity I am running is 3.8.3-1.el7

**Reply by nebfield** [reply, marked as answer]
Could you update the version of lmod installed on your cluster?

https://github.com/TACC/Lmod/issues/384

Or try overwriting the `__lmod_vx` environment variable as described here:

https://github.com/NOAA-EMC/hpc-stack/issues/54#issuecomment-845268971

**Reply by drzhangyin** [reply]
I will contact the cluster administrators to see if they can do so. Thank you~

**Reply by drzhangyin** [reply]
Hello and thanks again! 
I retested $ nextflow run pgscatalog/pgsc_calc -profile test,singularity  on a new cluster partition that with the latest version of lmod installed, and it ran with no error message. The output log file said "Pipeline completed successfully-". However, I didn't see score/ or match/ in the results directory, and the only folder I can see there is pipeline_info. Not sure is this normal?

I then tried to calculate  PGS000790 and got the error message below. Is this solely due to issues with PGS Catalog API or could there be other reasons?

Command exit status:
  1

Command output:
  Error Connecting: HTTPSConnectionPool(host='www.pgscatalog.org', port=443): Max retries exceeded with url: /rest/score/search?pgs_ids=PGS000790 (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x2b78bb450700>: Failed to establish a new connection: [Errno 101] Network is unreachable'))

Command error:
  bash: line 15: __lmod_vx: unbound variable
  root: 2023-05-16 21:03:46 DEBUG    Verbose logging enabled
  pgscatalog_utils.download.download_scorefile: 2023-05-16 21:03:46 CRITICAL Downloading harmonized scoring file(s) in build: GRCh37.
  pgscatalog_utils.download.download_scorefile: 2023-05-16 21:03:46 DEBUG    --id set, querying scores
  urllib3.connectionpool: 2023-05-16 21:03:46 DEBUG    Starting new HTTPS connection (1): www.pgscatalog.org:443
  pgscatalog_utils.download.score: 2023-05-16 21:05:53 WARNING  No results in response from PGS Catalog API. Please check the PGS IDs.
  pgscatalog_utils.download.score: 2023-05-16 21:05:53 ERROR    Bad response from PGS Catalog API. Is ['PGS000790'] a valid ID?
  Error Connecting: HTTPSConnectionPool(host='www.pgscatalog.org', port=443): Max retries exceeded with url: /rest/score/search?pgs_ids=PGS000790 (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x2b78bb450700>: Failed to establish a new connection: [Errno 101] Network is unreachable'))

**Reply by nebfield** [reply]
The results directory isn't normal. Running v1.3.2 on the test profile should output the following structure:

```
$ tree results/
results
├── match
│   ├── cineca-synthetic-subset_22_additive_0.scorefile.gz
│   ├── cineca-synthetic-subset_log.csv.gz
│   ├── cineca-synthetic-subset_summary.csv
│   └── versions.yml
├── pipeline_info
│   ├── execution_report_2023-05-17_08-48-13.html
│   ├── execution_timeline_2023-05-17_08-48-13.html
│   ├── execution_trace_2023-05-17_08-48-13.txt
│   └── pipeline_dag_2023-05-17_08-48-13.html
└── score
    ├── aggregated_scores.txt.gz
    ├── report.html
    └── versions.yml
```

You could try changing the output directory (`--outdir`) to a different place, maybe `$HOME/results`:

```
$ nextflow run pgscatalog/pgsc_calc -profile test,singularity --outdir $HOME/results
```

You could test re-running the workflow with the `-resume` option to see if your network problem was intermittent. The PGS Catalog API is OK:

```
$ curl -X GET "http://www.pgscatalog.org/rest/score/search?pgs_ids=PGS000790"
{"size":1,"count":1,"next":null,"previous":null,"results":[{"id":"PGS000790","name":"CC_Melanoma_IV","ftp_scoring_file":"https://ftp.ebi.ac.uk/pub/databases/spot/pgs/scores/PGS000790/ScoringFiles/PGS000790.txt.gz","ftp_harmonized_scoring_files":{"GRCh37":{"positions":"https://ftp.ebi.ac.uk/pub/databases/spot/pgs/scores/PGS000790/ScoringFiles/Harmonized/PGS000790_hmPOS_GRCh37.txt.gz"},"GRCh38":{"positions":"https://ftp.ebi.ac.uk/pub/databases/spot/pgs/scores/PGS000790/ScoringFiles/Harmonized/PGS000790_hmPOS_GRCh38.txt.gz"}},"publication":{"id":"PGP000186","title":"Pan-cancer analysis demonstrates that integrating polygenic risk scores with modifiable risk factors improves risk prediction.","doi":"10.1038/s41467-020-19600-4","PMID":33247094,"journal":"Nat Commun","firstauthor":"Kachuri L","date_publication":"2020-11-27"},"matches_publication":true,"samples_variants":[{"sample_number":2830,"sample_cases":null,"sample_controls":null,"sample_percent_male":null,"sample_age":null,"phenotyping_free":null,"followup_time":null,"ancestry_broad":"European","ancestry_free":null,"ancestry_country":"U.S.","ancestry_additional":null,"source_GWAS_catalog":"GCST001245","source_PMID":21926416,"source_DOI":null,"cohorts":[{"name_short":"MDACCS","name_full":"MD Anderson Cancer Center Study","name_others":null}],"cohorts_additional":null},{"sample_number":1728,"sample_cases":null,"sample_controls":null,"sample_percent_male":null,"sample_age":null,"phenotyping_free":null,"followup_time":null,"ancestry_broad":"European","ancestry_free":null,"ancestry_country":"Australia","ancestry_additional":null,"source_GWAS_catalog":"GCST000198","source_PMID":18488026,"source_DOI":null,"cohorts":[{"name_short":"Q-MEGA","name_full":"Queensland study of Melanoma: Environment and Genetic Associations","name_others":null}],"cohorts_additional":null},{"sample_number":36077,"sample_cases":null,"sample_controls":null,"sample_percent_male":null,"sample_age":null,"phenotyping_free":null,"followup_time":null,"ancestry_broad":"European","ancestry_free":null,"ancestry_country":"U.S., Israel, Australia, Germany, U.K., France","ancestry_additional":null,"source_GWAS_catalog":"GCST003061","source_PMID":26237428,"source_DOI":null,"cohorts":[],"cohorts_additional":null},{"sample_number":6555,"sample_cases":null,"sample_controls":null,"sample_percent_male":null,"sample_age":null,"phenotyping_free":null,"followup_time":null,"ancestry_broad":"European","ancestry_free":null,"ancestry_country":"Australia","ancestry_additional":null,"source_GWAS_catalog":"GCST001266","source_PMID":21983785,"source_DOI":null,"cohorts":[{"name_short":"QIMR","name_full":"Queensland Institute of Medical Research","name_others":null},{"name_short":"BATS","name_full":"Brisbane Adolescent Twins Study","name_others":null},{"name_short":"AMFS","name_full":"Australian Melanoma Family Study","name_others":null},{"name_short":"Q-MEGA","name_full":"Queensland study of Melanoma: Environment and Genetic Associations","name_others":null}],"cohorts_additional":null},{"sample_number":291407,"sample_cases":null,"sample_controls":null,"sample_percent_male":null,"sample_age":null,"phenotyping_free":null,"followup_time":null,"ancestry_broad":"European","ancestry_free":null,"ancestry_country":"U.S.","ancestry_additional":null,"source_GWAS_catalog":"GCST004142","source_PMID":28212542,"source_DOI":null,"cohorts":[{"name_short":"23andMe","name_full":"23andMe","name_others":null}],"cohorts_additional":null},{"sample_number":6122,"sample_cases":null,"sample_controls":null,"sample_percent_male":null,"sample_age":null,"phenotyping_free":null,"followup_time":null,"ancestry_broad":"European","ancestry_free":null,"ancestry_country":"U.S.","ancestry_additional":null,"source_GWAS_catalog":"GCST002514","source_PMID":24980573,"source_DOI":null,"cohorts":[{"name_short":"NHS","name_full":"Nurses Health Study","name_others":null},{"name_short":"HPFS","name_full":"Health Professional Follow-up Study","name_others":null}],"cohorts_additional":null},{"sample_number":10422,"sample_cases":null,"sample_controls":null,"sample_percent_male":null,"sample_age":null,"phenotyping_free":null,"followup_time":null,"ancestry_broad":"European","ancestry_free":null,"ancestry_country":"Sweden,U.S.,Poland,Israel,Australia,Italy,Netherlands,U.K.,Spain,France,Norway","ancestry_additional":null,"source_GWAS_catalog":"GCST001267","source_PMID":21983787,"source_DOI":null,"cohorts":[{"name_short":"GenoMEL","name_full":"GenoMEL","name_others":null},{"name_short":"B58C","name_full":"British 1958 Birth Cohort Study","name_others":null},{"name_short":"UKBS","name_full":"UK National Blood Donor Service","name_others":null},{"name_short":"CNG","name_full":"Centre National de G_notypage","name_others":null}],"cohorts_additional":null},{"sample_number":5456,"sample_cases":null,"sample_controls":null,"sample_percent_male":null,"sample_age":null,"phenotyping_free":null,"followup_time":null,"ancestry_broad":"European","ancestry_free":null,"ancestry_country":"Sweden,Australia,Italy,Netherlands,U.K.,Spain,France","ancestry_additional":null,"source_GWAS_catalog":"GCST000437","source_PMID":19578364,"source_DOI":null,"cohorts":[{"name_short":"GenoMEL","name_full":"GenoMEL","name_others":null},{"name_short":"WTCCC","name_full":"Wellcome Trust Case Control Consortium","name_others":null},{"name_short":"CNG","name_full":"Centre National de G_notypage","name_others":null}],"cohorts_additional":null},{"sample_number":4919,"sample_cases":null,"sample_controls":null,"sample_percent_male":null,"sample_age":null,"phenotyping_free":null,"followup_time":null,"ancestry_broad":"European","ancestry_free":null,"ancestry_country":null,"ancestry_additional":null,"source_GWAS_catalog":"GCST001886","source_PMID":23455637,"source_DOI":null,"cohorts":[{"name_short":"GenoMEL","name_full":"GenoMEL","name_others":null},{"name_short":"WTCCC","name_full":"Wellcome Trust Case Control Consortium","name_others":null}],"cohorts_additional":null}],"samples_training":[],"trait_reported":"Melanoma","trait_additional":null,"trait_efo":[{"id":"EFO_0000756","label":"melanoma","description":"A malignant, usually aggressive tumor composed of atypical, neoplastic melanocytes. Most often, melanomas arise in the skin (cutaneous melanomas) and include the following histologic subtypes: superficial spreading melanoma, nodular melanoma, acral lentiginous melanoma, and lentigo maligna melanoma. Cutaneous melanomas may arise from acquired or congenital melanocytic or dysplastic nevi. Melanomas may also arise in other anatomic sites including the gastrointestinal system, eye, urinary tract, and reproductive system. Melanomas frequently metastasize to lymph nodes, liver, lungs, and brain. [NCIT: C3224]","url":"http://www.ebi.ac.uk/efo/EFO_0000756"}],"method_name":"24 variants from Graff et al (PGS000079) with inverse variant weights","method_params":"NR","variants_number":24,"variants_interactions":0,"variants_genomebuild":"GRCh37","weight_type":"Inverse-variance weighting","ancestry_distribution":{"eval":{"dist":{"EUR":100},"count":1},"gwas":{"dist":{"EUR":100},"count":365516}},"date_release":"2021-05-28","license":"PGS obtained from the Catalog should be cited appropriately, and used in accordance with any licensing restrictions set by the authors. See EBI Terms of Use (https://www.ebi.ac.uk/about/terms-of-use/) for additional details."}]}%
```

If you have consistent problems with the PGS Catalog API, maybe your system administrator could help to fix your network problems.

I hope that helps 😄

**Reply by drzhangyin** [reply]
Thanks, and this is very helpful. :smile:
Just a quick update that I retested $ nextflow run pgscatalog/pgsc_calc -profile test,singularity using another output directory and it now works. 
The network problem seems to be consistent, so I am contacting the cluster administrators to see if they can help me fix it~

**Reply by nebfield** [reply]
Great! As a temporary workaround you could manually download the score from the PGS Catalog FTP in the correct genome build:

https://ftp.ebi.ac.uk/pub/databases/spot/pgs/scores/PGS000790/ScoringFiles/Harmonized/

And run `pgsc_calc` using the `--scorefile` parameter, e.g.:

```
$ nextflow run pgscatalog/pgsc_calc -profile test,singularity --outdir $HOME/results --scorefile path/to/scorefile.txt.gz
```

This workflow should run, but PGS000790 won't be calculated on the test profile because the test genomes are small and synthetic. It should work on your own data OK.

**Reply by drzhangyin** [reply]
Thanks!! I just tested $ nextflow run pgscatalog/pgsc_calc -profile test,singularity --outdir $HOME/results --scorefile path/to/scorefile.txt.gz and it runs successfully with no error! 

Based on this approach, I then tested PGS000004 by manually downloading the score from the PGS Catalog FTP and use the --scorefile parameter (please see below, not sure is it correct? :sweat_smile:):

./nextflow run pgscatalog/pgsc_calc \
    -profile singularity \
    --input samplesheet.csv --target_build GRCh37 \
    --scorefile XX/XX/pgscatalog_pipeline_weights_files_gz/breast_PGS000004  (the path to scorefile.txt.gz)

However, I got the error message below: 

Staging foreign file: https://gitlab.ebi.ac.uk/nebfield/test-datasets/-/raw/master/pgsc_calc/reference_data/pgsc_calc_ref.sqlar
ERROR ~ Error executing process > 'PGSCATALOG_PGSCALC:PGSCALC:INPUT_CHECK:COMBINE_SCOREFILES (1)'

Caused by:
  Process `PGSCATALOG_PGSCALC:PGSCALC:INPUT_CHECK:COMBINE_SCOREFILES (1)` terminated with an error exit status (1)

Command executed:

  combine_scorefiles -s breast_PGS000004             -t GRCh37             -o scorefiles.txt.gz             -l log_scorefiles.json             -v
  
  cat <<-END_VERSIONS > versions.yml
  COMBINE_SCOREFILES:
      pgscatalog_utils: $(echo $(python -c 'import pgscatalog_utils; print(pgscatalog_utils.__version__)'))
  END_VERSIONS

Command exit status:
  1

Command output:
  (empty)

Command error:
  bash: line 15: __lmod_vx: unbound variable
  root: 2023-05-18 12:55:24 DEBUG    Verbose logging enabled
  pgscatalog_utils.scorefile.combine_scorefiles: 2023-05-18 12:55:24 DEBUG    Input scorefiles: ['breast_PGS000004']
  pgscatalog_utils.scorefile.read: 2023-05-18 12:55:24 DEBUG    Reading scorefile breast_PGS000004
  Traceback (most recent call last):
    File "/venv/bin/combine_scorefiles", line 8, in <module>
      sys.exit(combine_scorefiles())
    File "/venv/lib/python3.10/site-packages/pgscatalog_utils/scorefile/combine_scorefiles.py", line 62, in combine_scorefiles
      h, score = load_scorefile(x)
    File "/venv/lib/python3.10/site-packages/pgscatalog_utils/scorefile/read.py", line 13, in load_scorefile
      df = pd.read_table(path, dtype=_scorefile_dtypes(), comment='#', na_values=['None'], low_memory=False)
    File "/venv/lib/python3.10/site-packages/pandas/util/_decorators.py", line 211, in wrapper
      return func(*args, **kwargs)
    File "/venv/lib/python3.10/site-packages/pandas/util/_decorators.py", line 317, in wrapper
      return func(*args, **kwargs)
    File "/venv/lib/python3.10/site-packages/pandas/io/parsers/readers.py", line 1289, in read_table
      return _read(filepath_or_buffer, kwds)
    File "/venv/lib/python3.10/site-packages/pandas/io/parsers/readers.py", line 605, in _read
      parser = TextFileReader(filepath_or_buffer, **kwds)
    File "/venv/lib/python3.10/site-packages/pandas/io/parsers/readers.py", line 1442, in __init__
      self._engine = self._make_engine(f, self.engine)
    File "/venv/lib/python3.10/site-packages/pandas/io/parsers/readers.py", line 1729, in _make_engine
      self.handles = get_handle(
    File "/venv/lib/python3.10/site-packages/pandas/io/common.py", line 857, in get_handle
      handle = open(
  IsADirectoryError: [Errno 21] Is a directory: 'breast_PGS000004'

**Reply by nebfield** [reply]
`--scorefile` has to be a path to a file, but it looks like you set the parameter to a directory. Try something like:

```
--scorefile XX/XX/pgscatalog_pipeline_weights_files_gz/breast_PGS000004/PGS000004_hmPOS_GRCh38.txt.gz
```

I'm assuming you download the scoring file from:

https://ftp.ebi.ac.uk/pub/databases/spot/pgs/scores/PGS000004/ScoringFiles/Harmonized/

And your genomes are in build GRCh38. Please replace `PGS000004_hmPOS_GRCh38.txt.gz` with the name of the file you download.

**Reply by drzhangyin** [reply]
Great catch! Just retested it based on your suggestion, and it now works. Thank you SO MUCH! :thumbsup:
Just one quick question: from the output summary report, I see among the 313 variants included in PGS000004, the matching rate is 87.9% and 30 ambiguous variants that were excluded. Just curious for those ambiguous variants, is there any "flip" option other than exclude in our pipeline?

**Reply by nebfield** [reply]
Great! 🥳

The `--keep_ambiguous` parameter [might be helpful](https://pgsc-calc.readthedocs.io/en/latest/reference/params.html#variant-matching-qc-filtering)

**Reply by drzhangyin** [reply]
Just tried it as below, and I got an error message "--keep_ambiguous: command not found". Did I put --keep_ambiguous in the wrong place?
 
./nextflow run pgscatalog/pgsc_calc \
    -profile singularity \
    --input samplesheet.csv --target_build GRCh37 \
    --scorefile XX/XX/pgscatalog_pipeline_weights_files_gz/breast_PGS000004/PGS000004_hmPOS_GRCh37.txt.gz
    --keep_ambiguous   

Also another quick question: Is there any option that allows us to see which specific variants are failed in matching in addition to the summary statistics showed on the summary report .html?  

Appreciate you time!

**Reply by nebfield** [reply]
Try putting everything on a single line in the terminal, like: 

```
$ nextflow run pgscatalog/pgsc_calc -profile singularity --input samplesheet.csv --target_build GRCh37 --scorefile XX/XX/pgscatalog_pipeline_weights_files_gz/breast_PGS000004/PGS000004_hmPOS_GRCh37.txt.gz --keep_ambiguous
```

A full variant matching log is available in `results/match`:

```
$ tree results/match
results/match
├── cineca-synthetic-subset_22_additive_0.scorefile.gz
├── cineca-synthetic-subset_log.csv.gz                               <--- this file here 
├── cineca-synthetic-subset_summary.csv
└── versions.yml
```

(these files are made from the test profile, your files will have different names)

**Reply by drzhangyin** [reply]
Wonderful! Thank you~ :thumbsup:
